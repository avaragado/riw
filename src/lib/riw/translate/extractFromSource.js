// @flow

import { transformFileSync, type babel } from 'babel-core';
import pluginReactIntl from 'babel-plugin-react-intl';
import glob from 'glob';
import chain from 'ramda/src/chain';
import pathOr from 'ramda/src/pathOr';
import compose from 'ramda/src/compose';

const arfabsFromConfig: FilesFromConfig = config => chain(
    (sGlob: Glob) => glob.sync(sGlob, { cwd: config.rootDir, absolute: true }),
    config.sourceDirs,
);

const outputBabelFromFabs = (fabs: AbsolutePath): babel.BabelFileResult => {
    try {
        return transformFileSync(fabs, {
            plugins: [pluginReactIntl],
        });

    } catch (err) {
        console.log(err);
        throw err;
    }
};

const armdFromFabs: MessageDescriptorsFromFile = compose(
    pathOr([], ['metadata', 'react-intl', 'messages']),
    outputBabelFromFabs,
);

export {
    arfabsFromConfig,
    armdFromFabs,
};
