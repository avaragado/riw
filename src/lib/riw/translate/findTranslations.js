// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import reduce from 'ramda/src/reduce';
import pathOr from 'ramda/src/pathOr';
import assocPath from 'ramda/src/assocPath';

import { sDescriptionDefault } from '../';

type LocaleMapper = (config: RIWConfig) => (md: RIWMessageDescriptor) =>
    RIWMessageDescriptorUntranslated;

type TranslationLookup = {
    mdu: RIWMessageDescriptorUntranslated,
    result: RIWDBTranslatedMessage | null,
};

type LookerUpper = (config: RIWConfig, db: RIWDB) => (mdu: RIWMessageDescriptorUntranslated) =>
    TranslationLookup;

type LookupReducer = (acc: RIWFindTranslationResult, lookup: TranslationLookup) =>
    RIWFindTranslationResult;

type TranslationFinder = <T>(
    config: RIWConfig,
    notify: string => T => T,
    db: RIWDB
) => (armd: RIWMessageDescriptor[]) => RIWFindTranslationResult;


const resultEmpty: RIWFindTranslationResult = {
    locale: {},
    armdu: [],
};

const eachLocale: LocaleMapper = config => md => map(
    locale => ({ ...md, locale }),
    config.targetLocales.concat(config.defaultLocale),
);

const includeLookup: LookerUpper = (config, db) => mdu => ({
    mdu,
    result: mdu.locale === config.defaultLocale
        ? mdu.defaultMessage
        : pathOr(null, [
            'data',
            mdu.defaultMessage,
            mdu.description || sDescriptionDefault,
            mdu.locale,
        ], db),
});

const reduceLookup: LookupReducer = (acc, lookup) => {
    const { mdu, result } = lookup;

    if (result) {
        return assocPath(['locale', mdu.locale, mdu.id], result, acc);
    }

    return {
        ...acc,
        armdu: acc.armdu.concat(mdu),
    };
};

const find: TranslationFinder = (config, notify, db) => compose(
    notify('endLookup'),
    reduce(reduceLookup, resultEmpty),
    map(includeLookup(config, db)),
    chain(eachLocale(config)),
    notify('startLookup'),
);

export default find;
