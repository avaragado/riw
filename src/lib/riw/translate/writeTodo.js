// @flow

import fs from 'fs';
import path from 'path';

import mkdirp from 'mkdirp';

import log from '../../../lib/log';

const stringify = obj => JSON.stringify(obj, null, 4);

export default (config: RIWConfig, notify: string => string => string) =>
    (translation: RIWFindTranslationResult) => {
        const { todoFile: pathOut, rootDir } = config;

        const fabsOut = path.resolve(rootDir || process.cwd(), pathOut);
        const dabsOut = path.dirname(fabsOut);

        try {
            mkdirp.sync(dabsOut);
        } catch (err) {
            log.error('riw', `Unable to create directory ${dabsOut} for riw todo file(s).`);
            throw err;
        }

        if (config.outputMode !== 'no-file') {
            fs.writeFileSync(fabsOut, stringify(translation.armdu));
            notify('fileSaved')(fabsOut);
        }

        return translation;
    };
