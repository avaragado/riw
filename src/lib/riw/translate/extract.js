// @flow

import chain from 'ramda/src/chain';
import compose from 'ramda/src/compose';
import map from 'ramda/src/map';
import prop from 'ramda/src/prop';

import * as source from './extractFromSource';
import * as json from './extractFromJSON';

type MDExtractor = (arfabsFromConfig: FilesFromConfig, armdFromFabs: MessageDescriptorsFromFile) =>
    <T>(notify: string => T => T) =>
    (config: RIWConfig) => RIWMessageDescriptor[];

const armdExtract: MDExtractor = (arfabsFromConfig, armdFromFabs) => notify => compose(
    notify('endExtract'),
    chain(prop('armd')),
    map(compose(
        notify('endExtractFile'),
        fabs => ({
            fabs,
            armd: map(
                md => ({ ...md, fabs }),
                armdFromFabs(fabs),
            ),
        }),
        notify('startExtractFile'),
    )),
    notify('startExtract'),
    arfabsFromConfig,
);

export const armdExtractSource = armdExtract(source.arfabsFromConfig, source.armdFromFabs);
export const armdExtractJSON = armdExtract(json.arfabsFromConfig, json.armdFromFabs);
