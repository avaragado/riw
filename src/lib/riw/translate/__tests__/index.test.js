// @flow

import path from 'path';
import fs from 'fs';

import mock from 'mock-fs';
import outdent from 'outdent';

import translate from '../';
import cfgBase from '../../__tests__/helpers/dummyConfig';

const absify = (name, frel) => path.resolve('fixtures', name, frel);
const stringify = obj => JSON.stringify(obj, null, 4);

const db: RIWDB = {
    version: 1,
    data: {
        'a.1!': {
            '_': {
                'en-up': 'A.1!',
                'en-ne': '!1.a',
            },
        },
        'b.1!': {
            '_': {
                'en-ne': '!1.b',
            },
        },
        'b.2!': {
            'b.2 desc': {
                'en-up': 'B.2!',
            },
        },
    },
};

const cfgOverride = {
    ...cfgBase,
    rootDir: '.',
    defaultLocale: 'en-us',
    targetLocales: ['en-up', 'en-ne'],
    translationsDatabaseFile: 'fixtures/db.json',
    translationsOutputFile: 'fixtures/locale/[locale].json',
    outputMode: 'file-per-locale',
    todoFile: 'fixtures/locale/todo.json',
};

type Fixture = {
    name: string,
    dir: { [key: string]: string },
    configOverride?: Object,
    out: RIWCLIProjectTranslateResult,
    fs: Object,
};

const fixtures: Fixture[] = [
    {
        name: '01',
        dir: {}, // empty
        out: {
            armd: [],
            dups: [],
            translation: {
                locale: {},
                armdu: [],
            },
        },
        fs: {
            'en-us.json': {},
            'en-up.json': {},
            'en-ne.json': {},
            'todo.json': [],
        },
    },

    {
        name: '02',
        dir: {
            'a.js': 'export default a => a;',
            'b.js': 'export default b => b + 1;',
        },
        out: {
            armd: [],
            dups: [],
            translation: {
                locale: {},
                armdu: [],
            },
        },
        fs: {
            'en-us.json': {},
            'en-up.json': {},
            'en-ne.json': {},
            'todo.json': [],
        },
    },

    {
        name: '03',
        dir: {
            'a.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    a1: {
                        id: 'a.1',
                        defaultMessage: 'a.1!',
                    },
                });
            `,
            'aa.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    a1again: {
                        id: 'a.1',
                        defaultMessage: 'a.1 again!',
                    },
                });
            `,
            'b.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    b1: {
                        id: 'b.1',
                        defaultMessage: 'b.1!',
                    },
                    b2: {
                        id: 'b.2',
                        description: 'b.2 desc',
                        defaultMessage: 'b.2!',
                    },
                });
            `,
            'bb.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    b2again: {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                    },
                });
            `,
            'c.js': outdent`
                import { defineMessages } from 'react-intl';

                const msg = defineMessages({
                    c1: {
                        id: 'c.1',
                        defaultMessage: 'c.1!',
                    },
                });
            `,
        },
        out: {
            armd: [
                {
                    id: 'a.1',
                    description: undefined,
                    defaultMessage: 'a.1!',
                    fabs: absify('03', 'a.js'),
                },
                {
                    id: 'a.1',
                    description: undefined,
                    defaultMessage: 'a.1 again!',
                    fabs: absify('03', 'aa.js'),
                },
                {
                    id: 'b.1',
                    description: undefined,
                    defaultMessage: 'b.1!',
                    fabs: absify('03', 'b.js'),
                },
                {
                    id: 'b.2',
                    description: 'b.2 desc',
                    defaultMessage: 'b.2!',
                    fabs: absify('03', 'b.js'),
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('03', 'bb.js'),
                },
                {
                    id: 'c.1',
                    description: undefined,
                    defaultMessage: 'c.1!',
                    fabs: absify('03', 'c.js'),
                },
            ],
            dups: [
                {
                    id: 'a.1',
                    arfabs: ['a.js', 'aa.js'].map(frel => absify('03', frel)),
                },
                {
                    id: 'b.2',
                    arfabs: ['b.js', 'bb.js'].map(frel => absify('03', frel)),
                },
            ],
            translation: {
                locale: {
                    'en-us': {
                        'a.1': 'a.1 again!',
                        'b.1': 'b.1!',
                        'b.2': 'b.2 again!',
                        'c.1': 'c.1!',
                    },
                    'en-up': {
                        'a.1': 'A.1!',
                        'b.2': 'B.2!',
                    },
                    'en-ne': {
                        'a.1': '!1.a',
                        'b.1': '!1.b',
                    },
                },
                armdu: [
                    {
                        id: 'a.1',
                        description: undefined,
                        defaultMessage: 'a.1 again!',
                        fabs: absify('03', 'aa.js'),
                        locale: 'en-up',
                    },
                    {
                        id: 'a.1',
                        description: undefined,
                        defaultMessage: 'a.1 again!',
                        fabs: absify('03', 'aa.js'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'b.1',
                        description: undefined,
                        defaultMessage: 'b.1!',
                        fabs: absify('03', 'b.js'),
                        locale: 'en-up',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 desc',
                        defaultMessage: 'b.2!',
                        fabs: absify('03', 'b.js'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                        fabs: absify('03', 'bb.js'),
                        locale: 'en-up',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                        fabs: absify('03', 'bb.js'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'c.1',
                        description: undefined,
                        defaultMessage: 'c.1!',
                        fabs: absify('03', 'c.js'),
                        locale: 'en-up',
                    },
                    {
                        id: 'c.1',
                        description: undefined,
                        defaultMessage: 'c.1!',
                        fabs: absify('03', 'c.js'),
                        locale: 'en-ne',
                    },
                ],
            },
        },
        fs: {
            'en-us.json': {
                'a.1': 'a.1 again!',
                'b.1': 'b.1!',
                'b.2': 'b.2 again!',
                'c.1': 'c.1!',
            },
            'en-up.json': {
                'a.1': 'A.1!',
                'b.2': 'B.2!',
            },
            'en-ne.json': {
                'a.1': '!1.a',
                'b.1': '!1.b',
            },
            'todo.json': [
                {
                    id: 'a.1',
                    description: undefined,
                    defaultMessage: 'a.1 again!',
                    fabs: absify('03', 'aa.js'),
                    locale: 'en-up',
                },
                {
                    id: 'a.1',
                    description: undefined,
                    defaultMessage: 'a.1 again!',
                    fabs: absify('03', 'aa.js'),
                    locale: 'en-ne',
                },
                {
                    id: 'b.1',
                    description: undefined,
                    defaultMessage: 'b.1!',
                    fabs: absify('03', 'b.js'),
                    locale: 'en-up',
                },
                {
                    id: 'b.2',
                    description: 'b.2 desc',
                    defaultMessage: 'b.2!',
                    fabs: absify('03', 'b.js'),
                    locale: 'en-ne',
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('03', 'bb.js'),
                    locale: 'en-up',
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('03', 'bb.js'),
                    locale: 'en-ne',
                },
                {
                    id: 'c.1',
                    description: undefined,
                    defaultMessage: 'c.1!',
                    fabs: absify('03', 'c.js'),
                    locale: 'en-up',
                },
                {
                    id: 'c.1',
                    description: undefined,
                    defaultMessage: 'c.1!',
                    fabs: absify('03', 'c.js'),
                    locale: 'en-ne',
                },
            ],
        },
    },

    {
        name: '04',
        dir: {
            'a.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1!"
                    }
                ]
            `,
            'aa.json': outdent`
                [
                    {
                        "id": "a.1",
                        "defaultMessage": "a.1 again!"
                    }
                ]
            `,
            'b.json': outdent`
                [
                    {
                        "id": "b.1",
                        "defaultMessage": "b.1!"
                    },
                    {
                        "id": "b.2",
                        "description": "b.2 desc",
                        "defaultMessage": "b.2!"
                    }
                ]
            `,
            'bb.json': outdent`
                [
                    {
                        "id": "b.2",
                        "description": "b.2 again desc",
                        "defaultMessage": "b.2 again!"
                    }
                ]
            `,
            'c.json': outdent`
                [
                    {
                        "id": "c.1",
                        "defaultMessage": "c.1!"
                    }
                ]
            `,
        },
        configOverride: {
            collateDir: 'fixtures/04',
            inputMode: 'json',
        },
        out: {
            armd: [
                {
                    id: 'a.1',
                    defaultMessage: 'a.1!',
                    fabs: absify('04', 'a.json'),
                },
                {
                    id: 'a.1',
                    defaultMessage: 'a.1 again!',
                    fabs: absify('04', 'aa.json'),
                },
                {
                    id: 'b.1',
                    defaultMessage: 'b.1!',
                    fabs: absify('04', 'b.json'),
                },
                {
                    id: 'b.2',
                    description: 'b.2 desc',
                    defaultMessage: 'b.2!',
                    fabs: absify('04', 'b.json'),
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('04', 'bb.json'),
                },
                {
                    id: 'c.1',
                    defaultMessage: 'c.1!',
                    fabs: absify('04', 'c.json'),
                },
            ],
            dups: [
                {
                    id: 'a.1',
                    arfabs: ['a.json', 'aa.json'].map(frel => absify('04', frel)),
                },
                {
                    id: 'b.2',
                    arfabs: ['b.json', 'bb.json'].map(frel => absify('04', frel)),
                },
            ],
            translation: {
                locale: {
                    'en-us': {
                        'a.1': 'a.1 again!',
                        'b.1': 'b.1!',
                        'b.2': 'b.2 again!',
                        'c.1': 'c.1!',
                    },
                    'en-up': {
                        'a.1': 'A.1!',
                        'b.2': 'B.2!',
                    },
                    'en-ne': {
                        'a.1': '!1.a',
                        'b.1': '!1.b',
                    },
                },
                armdu: [
                    {
                        id: 'a.1',
                        defaultMessage: 'a.1 again!',
                        fabs: absify('04', 'aa.json'),
                        locale: 'en-up',
                    },
                    {
                        id: 'a.1',
                        defaultMessage: 'a.1 again!',
                        fabs: absify('04', 'aa.json'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'b.1',
                        defaultMessage: 'b.1!',
                        fabs: absify('04', 'b.json'),
                        locale: 'en-up',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 desc',
                        defaultMessage: 'b.2!',
                        fabs: absify('04', 'b.json'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                        fabs: absify('04', 'bb.json'),
                        locale: 'en-up',
                    },
                    {
                        id: 'b.2',
                        description: 'b.2 again desc',
                        defaultMessage: 'b.2 again!',
                        fabs: absify('04', 'bb.json'),
                        locale: 'en-ne',
                    },
                    {
                        id: 'c.1',
                        defaultMessage: 'c.1!',
                        fabs: absify('04', 'c.json'),
                        locale: 'en-up',
                    },
                    {
                        id: 'c.1',
                        defaultMessage: 'c.1!',
                        fabs: absify('04', 'c.json'),
                        locale: 'en-ne',
                    },
                ],
            },
        },
        fs: {
            'en-us.json': {
                'a.1': 'a.1 again!',
                'b.1': 'b.1!',
                'b.2': 'b.2 again!',
                'c.1': 'c.1!',
            },
            'en-up.json': {
                'a.1': 'A.1!',
                'b.2': 'B.2!',
            },
            'en-ne.json': {
                'a.1': '!1.a',
                'b.1': '!1.b',
            },
            'todo.json': [
                {
                    id: 'a.1',
                    defaultMessage: 'a.1 again!',
                    fabs: absify('04', 'aa.json'),
                    locale: 'en-up',
                },
                {
                    id: 'a.1',
                    defaultMessage: 'a.1 again!',
                    fabs: absify('04', 'aa.json'),
                    locale: 'en-ne',
                },
                {
                    id: 'b.1',
                    defaultMessage: 'b.1!',
                    fabs: absify('04', 'b.json'),
                    locale: 'en-up',
                },
                {
                    id: 'b.2',
                    description: 'b.2 desc',
                    defaultMessage: 'b.2!',
                    fabs: absify('04', 'b.json'),
                    locale: 'en-ne',
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('04', 'bb.json'),
                    locale: 'en-up',
                },
                {
                    id: 'b.2',
                    description: 'b.2 again desc',
                    defaultMessage: 'b.2 again!',
                    fabs: absify('04', 'bb.json'),
                    locale: 'en-ne',
                },
                {
                    id: 'c.1',
                    defaultMessage: 'c.1!',
                    fabs: absify('04', 'c.json'),
                    locale: 'en-up',
                },
                {
                    id: 'c.1',
                    defaultMessage: 'c.1!',
                    fabs: absify('04', 'c.json'),
                    locale: 'en-ne',
                },
            ],
        },
    },
];

describe('lib/riw/translate', () => {
    afterEach(() => {
        mock.restore();
    });

    fixtures.forEach((fixture) => {
        it(fixture.name, () => {
            mock({
                fixtures: {
                    [fixture.name]: fixture.dir,
                    'db.json': stringify(db),
                    'locale': {},
                },
            });

            const cfg = {
                ...cfgOverride,
                sourceDirs: [`fixtures/${fixture.name}/**/*.js`],
                ...fixture.configOverride,
            };

            const received = translate(cfg)({});

            expect(received).toEqual(fixture.out);

            const arfrel = fs.readdirSync('fixtures/locale');

            expect(arfrel).toEqual(Object.keys(fixture.fs).sort());

            arfrel.forEach((frel) => {
                const content = JSON.parse(fs.readFileSync(`fixtures/locale/${frel}`).toString());

                expect(content).toEqual(fixture.fs[frel]);
            });
        });
    });
});
