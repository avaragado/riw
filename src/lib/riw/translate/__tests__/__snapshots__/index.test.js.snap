// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/riw/translate 01 1`] = `
Object {
  "armd": Array [],
  "dups": Array [],
  "translation": Object {
    "armdu": Array [],
    "locale": Object {},
  },
}
`;

exports[`lib/riw/translate 01 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/translate 01 3`] = `
Array [
  Object {},
  Object {},
  Object {},
  Array [],
]
`;

exports[`lib/riw/translate 02 1`] = `
Object {
  "armd": Array [],
  "dups": Array [],
  "translation": Object {
    "armdu": Array [],
    "locale": Object {},
  },
}
`;

exports[`lib/riw/translate 02 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/translate 02 3`] = `
Array [
  Object {},
  Object {},
  Object {},
  Array [],
]
`;

exports[`lib/riw/translate 03 1`] = `
Object {
  "armd": Array [
    Object {
      "defaultMessage": "a.1!",
      "description": undefined,
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.js",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "description": undefined,
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "b.1!",
      "description": undefined,
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
      "id": "b.1",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "c.1!",
      "description": undefined,
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.js",
      "id": "c.1",
    },
  ],
  "dups": Array [
    Object {
      "arfabs": Array [
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.js",
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
      ],
      "id": "a.1",
    },
    Object {
      "arfabs": Array [
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
      ],
      "id": "b.2",
    },
  ],
  "translation": Object {
    "armdu": Array [
      Object {
        "defaultMessage": "a.1 again!",
        "description": undefined,
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
        "id": "a.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "a.1 again!",
        "description": undefined,
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
        "id": "a.1",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.1!",
        "description": undefined,
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
        "id": "b.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2!",
        "description": "b.2 desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
        "id": "b.2",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "c.1!",
        "description": undefined,
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.js",
        "id": "c.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "c.1!",
        "description": undefined,
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.js",
        "id": "c.1",
        "locale": "en-ne",
      },
    ],
    "locale": Object {
      "en-ne": Object {
        "a.1": "!1.a",
        "b.1": "!1.b",
      },
      "en-up": Object {
        "a.1": "A.1!",
        "b.2": "B.2!",
      },
      "en-us": Object {
        "a.1": "a.1 again!",
        "b.1": "b.1!",
        "b.2": "b.2 again!",
        "c.1": "c.1!",
      },
    },
  },
}
`;

exports[`lib/riw/translate 03 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/translate 03 3`] = `
Array [
  Object {
    "a.1": "!1.a",
    "b.1": "!1.b",
  },
  Object {
    "a.1": "A.1!",
    "b.2": "B.2!",
  },
  Object {
    "a.1": "a.1 again!",
    "b.1": "b.1!",
    "b.2": "b.2 again!",
    "c.1": "c.1!",
  },
  Array [
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
      "id": "a.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.js",
      "id": "a.1",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
      "id": "b.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.js",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
      "id": "b.2",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.js",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.js",
      "id": "c.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.js",
      "id": "c.1",
      "locale": "en-ne",
    },
  ],
]
`;

exports[`lib/riw/translate 04 1`] = `
Object {
  "armd": Array [
    Object {
      "defaultMessage": "a.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.json",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
      "id": "a.1",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
      "id": "b.1",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
      "id": "b.2",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
      "id": "c.1",
    },
  ],
  "dups": Array [
    Object {
      "arfabs": Array [
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/a.json",
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
      ],
      "id": "a.1",
    },
    Object {
      "arfabs": Array [
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
        "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
      ],
      "id": "b.2",
    },
  ],
  "translation": Object {
    "armdu": Array [
      Object {
        "defaultMessage": "a.1 again!",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
        "id": "a.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "a.1 again!",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
        "id": "a.1",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.1!",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
        "id": "b.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2!",
        "description": "b.2 desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
        "id": "b.2",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "b.2 again!",
        "description": "b.2 again desc",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
        "id": "b.2",
        "locale": "en-ne",
      },
      Object {
        "defaultMessage": "c.1!",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
        "id": "c.1",
        "locale": "en-up",
      },
      Object {
        "defaultMessage": "c.1!",
        "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
        "id": "c.1",
        "locale": "en-ne",
      },
    ],
    "locale": Object {
      "en-ne": Object {
        "a.1": "!1.a",
        "b.1": "!1.b",
      },
      "en-up": Object {
        "a.1": "A.1!",
        "b.2": "B.2!",
      },
      "en-us": Object {
        "a.1": "a.1 again!",
        "b.1": "b.1!",
        "b.2": "b.2 again!",
        "c.1": "c.1!",
      },
    },
  },
}
`;

exports[`lib/riw/translate 04 2`] = `
Array [
  "en-ne.json",
  "en-up.json",
  "en-us.json",
  "todo.json",
]
`;

exports[`lib/riw/translate 04 3`] = `
Array [
  Object {
    "a.1": "!1.a",
    "b.1": "!1.b",
  },
  Object {
    "a.1": "A.1!",
    "b.2": "B.2!",
  },
  Object {
    "a.1": "a.1 again!",
    "b.1": "b.1!",
    "b.2": "b.2 again!",
    "c.1": "c.1!",
  },
  Array [
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
      "id": "a.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "a.1 again!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/aa.json",
      "id": "a.1",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
      "id": "b.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2!",
      "description": "b.2 desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/b.json",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
      "id": "b.2",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "b.2 again!",
      "description": "b.2 again desc",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/bb.json",
      "id": "b.2",
      "locale": "en-ne",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
      "id": "c.1",
      "locale": "en-up",
    },
    Object {
      "defaultMessage": "c.1!",
      "fabs": "/Users/dave/git-repos/react-intl-workflow/fixtures/dir/c.json",
      "id": "c.1",
      "locale": "en-ne",
    },
  ],
]
`;
